lib/apia/open_api/specification.rb

```diff
@@ -85,6 +96,55 @@ def add_security
         end
       end
 
+      def get_tag_group_index(tag)
+        @spec[:"x-tagGroups"].each_with_index do |group, index|
+          return index if !group.nil? && group[:name] == tag
+        end
+
+        nil
+      end
+
+      def add_tag_groups
+        @spec[:paths].each_value do |methods|
+          methods.each_value do |method_spec|
+            method_spec[:tags].each_with_index do |tag, tag_index|
+              unless @spec[:tags].any? { |t| t[:name] == tag }
+                @spec[:tags] << { name: tag }
+              end
+
+              next if tag_index.zero?
+
+              parent_tag = method_spec[:tags][tag_index - 1]
+              index = get_tag_group_index(parent_tag)
+
+              if index.nil? && method_spec[:tags].size > 1
+                @spec[:"x-tagGroups"] << { name: parent_tag, tags: [parent_tag] }
+                index = @spec[:"x-tagGroups"].size - 1
+              end
+
+              unless @spec[:"x-tagGroups"][index][:tags].include?(tag)
+                @spec[:"x-tagGroups"][index][:tags] << tag
+              end
+
+              index = get_tag_group_index(tag)
+
+              if index.nil? && method_spec[:tags].size > 1 && tag_index < method_spec[:tags].size - 1
+                @spec[:"x-tagGroups"] << { name: tag, tags: [] }
+              end
+            end
+          end
+        end
+
+        @spec[:tags].each do |tag|
+          unless @spec[:"x-tagGroups"].any? { |group| group[:tags].include?(tag[:name]) }
+            @spec[:"x-tagGroups"] << { name: tag[:name], tags: [tag[:name]] }
+          end
+        end
+
+        @spec[:"x-tagGroups"].sort_by! { |group| group[:name] }
+        @spec[:"x-tagGroups"].each { |group| group[:tags].sort! }
+      end
```

Comment:
I won't pretend to have verified the output here. But I'm assuming you know the format you want.

Reply:
I have tested this gives the desired output manually but I'll be honest this is the worst part of this PR. 
Anything that stands out as odd or bad in particular?